File: AuthorizationApiMicroservice/pkg/jwt/jwt.go
package jwt

import (
	"errors"
	"time"

	"github.com/golang-jwt/jwt/v4"
)

var (
	ErrInvalidToken = errors.New("invalid token")
	ErrExpiredToken = errors.New("token expired")
)

type Payload struct {
	Jti     string
	Subject string
	Iat     int64
	Exp     int64
}

func GenerateRS256Jwt(payload Payload, privateKey string) (string, error) {
	token := jwt.NewWithClaims(jwt.SigningMethodRS256, jwt.MapClaims{
		"jti":     payload.Jti,
		"subject": payload.Subject,
		"iat":     payload.Iat,
		"exp":     payload.Exp,
	})

	key, err := jwt.ParseRSAPrivateKeyFromPEM([]byte(privateKey))
	if err != nil {
		return "", err
	}

	tokenString, err := token.SignedString(key)
	if err != nil {
		return "", nil
	}

	return tokenString, nil
}

func ValidateRS256Jwt(tokenString string, publicKey string) (*Payload, error) {
	key, err := jwt.ParseRSAPublicKeyFromPEM([]byte(publicKey))
	if err != nil {
		return nil, err
	}

	token, err := jwt.ParseWithClaims(tokenString, jwt.MapClaims{}, func(t *jwt.Token) (interface{}, error) {
		return key, nil
	})

	if err != nil {
		return nil, err
	}

	payload, ok := token.Claims.(jwt.MapClaims)
	if !ok {
		return nil, ErrInvalidToken
	}

	exp := payload["exp"].(int64)
	if exp <= time.Now().Unix() {
		return nil, ErrExpiredToken
	}

	return &Payload{
		Jti:     payload["jti"].(string),
		Subject: payload["subject"].(string),
		Iat:     payload["iat"].(int64),
		Exp:     exp,
	}, nil
}

File: SsoMicroservice/src/libs/jwt.ts
import * as jwt from "jsonwebtoken"

export class JwtExpiredError extends Error {
  constructor() {
    super("Token expired")
  }
}

export class InvalidJwtError extends Error {
  constructor() {
    super("Invalid token")
  }
}

export type IJwtPayload = {
  jti: string
  sub: string
  iat: number
  exp: number
}

export const generateRS256Jwt = (payload: IJwtPayload, privateKey: string) => {
  return jwt.sign(payload, privateKey, { algorithm: "RS256" })
}

export const validateRS256Jwt = (token: string, publicKey: string) => {
  try {
    const payload = jwt.verify(token, publicKey)
    return payload as IJwtPayload
  } catch (err) {
    if (err instanceof jwt.TokenExpiredError) {
      throw new JwtExpiredError()
    } else {
      throw new InvalidJwtError()
    }
  }
}

export const decodeJwt = (token: string) => {
  try {
    const payload = jwt.decode(token)
    return payload as IJwtPayload
  } catch (err) {
    if (err instanceof jwt.JsonWebTokenError) {
      throw new InvalidJwtError()
    }
  }
}
